File: native/tet4d_core/src/core/topology_transport.cpp
#include "tet4d_core/topology_transport.hpp"

#include "tet4d_core/generated/topology_contract_v1.hpp"

#include <algorithm>
#include <cctype>
#include <set>
#include <sstream>
#include <utility>

namespace tet4d::core {
namespace {

template <typename T>
TopologyTransportResult<T> failure(
		std::string code,
		std::string path,
		std::string expected,
		std::string actual,
		std::string message) {
	TopologyTransportResult<T> result;
	result.error = TopologyTransportError{
			std::move(code),
			std::move(path),
			std::move(expected),
			std::move(actual),
			std::move(message)};
	return result;
}

template <typename T>
TopologyTransportResult<T> success(T value) {
	TopologyTransportResult<T> result;
	result.value = std::move(value);
	return result;
}

std::string trim(std::string value) {
	const auto first = std::find_if_not(value.begin(), value.end(), [](unsigned char ch) {
		return std::isspace(ch) != 0;
	});
	const auto last = std::find_if_not(value.rbegin(), value.rend(), [](unsigned char ch) {
		return std::isspace(ch) != 0;
	}).base();
	if (first >= last) {
		return {};
	}
	return std::string(first, last);
}

std::string lowercase(std::string value) {
	std::transform(value.begin(), value.end(), value.begin(), [](unsigned char ch) {
		return static_cast<char>(std::tolower(ch));
	});
	return value;
}

const TransportValue::Object *require_object(
		const TransportValue &value,
		const std::string &path,
		TopologyTransportError &error) {
	const auto *object = std::get_if<TransportValue::Object>(&value.value);
	if (object == nullptr) {
		error = {"wrong_type", path, "object", transport_value_category(value), path + " must be an object"};
	}
	return object;
}

const TransportValue::Array *require_array(
		const TransportValue &value,
		const std::string &path,
		TopologyTransportError &error) {
	const auto *array = std::get_if<TransportValue::Array>(&value.value);
	if (array == nullptr) {
		error = {"wrong_type", path, "array", transport_value_category(value), path + " must be an array"};
	}
	return array;
}

const TransportValue *required_field(
		const TransportValue::Object &object,
		const std::string &field,
		const std::string &path,
		TopologyTransportError &error) {
	const auto found = object.find(field);
	if (found == object.end()) {
		error = {"missing_field", path + "." + field, "present field", "missing", path + "." + field + " is required"};
		return nullptr;
	}
	return &found->second;
}

bool require_exact_fields(
		const TransportValue::Object &object,
		const std::set<std::string> &fields,
		const std::string &path,
		TopologyTransportError &error) {
	for (const std::string &field : fields) {
		if (object.find(field) == object.end()) {
			error = {"missing_field", path + "." + field, "present field", "missing", path + "." + field + " is required"};
			return false;
		}
	}
	for (const auto &[field, unused] : object) {
		(void)unused;
		if (fields.find(field) == fields.end()) {
			error = {"unknown_required_value", path + "." + field, "known field", "unknown field", path + "." + field + " is not supported"};
			return false;
		}
	}
	return true;
}

const std::int64_t *require_integer(
		const TransportValue &value,
		const std::string &path,
		TopologyTransportError &error) {
	const auto *integer = std::get_if<std::int64_t>(&value.value);
	if (integer == nullptr) {
		error = {"wrong_type", path, "integer", transport_value_category(value), path + " must be an integer"};
	}
	return integer;
}

const bool *require_boolean(
		const TransportValue &value,
		const std::string &path,
		TopologyTransportError &error) {
	const auto *boolean = std::get_if<bool>(&value.value);
	if (boolean == nullptr) {
		error = {"wrong_type", path, "boolean", transport_value_category(value), path + " must be a boolean"};
	}
	return boolean;
}

const std::string *require_string(
		const TransportValue &value,
		const std::string &path,
		TopologyTransportError &error) {
	const auto *string = std::get_if<std::string>(&value.value);
	if (string == nullptr) {
		error = {"wrong_type", path, "string", transport_value_category(value), path + " must be a string"};
	}
	return string;
}

template <typename Container, typename Value>
bool contains(const Container &values, const Value &value) {
	return std::find(values.begin(), values.end(), value) != values.end();
}

bool decode_integer_array(
		const TransportValue &value,
		const std::string &path,
		std::vector<std::int64_t> &result,
		TopologyTransportError &error) {
	const TransportValue::Array *array = require_array(value, path, error);
	if (array == nullptr) {
		return false;
	}
	result.clear();
	result.reserve(array->size());
	for (std::size_t index = 0; index < array->size(); ++index) {
		const std::string item_path = path + "[" + std::to_string(index) + "]";
		const std::int64_t *item = require_integer((*array)[index], item_path, error);
		if (item == nullptr) {
			return false;
		}
		result.push_back(*item);
	}
	return true;
}

bool decode_boundary(
		const TransportValue &value,
		const std::string &path,
		std::int64_t rank,
		TopologyTransportBoundary &result,
		TopologyTransportError &error) {
	const TransportValue::Object *object = require_object(value, path, error);
	if (object == nullptr || !require_exact_fields(*object, {"axis", "side"}, path, error)) {
		return false;
	}
	const TransportValue *axis_value = required_field(*object, "axis", path, error);
	const TransportValue *side_value = required_field(*object, "side", path, error);
	if (axis_value == nullptr || side_value == nullptr) {
		return false;
	}
	const std::string *raw_axis = require_string(*axis_value, path + ".axis", error);
	if (raw_axis == nullptr) {
		return false;
	}
	const std::string axis = lowercase(trim(*raw_axis));
	auto axis_found = std::find(generated::AXIS_NAMES.begin(), generated::AXIS_NAMES.end(), axis);
	if (axis_found == generated::AXIS_NAMES.end() ||
			std::distance(generated::AXIS_NAMES.begin(), axis_found) >= rank) {
		error = {"invalid_axis", path + ".axis", "axis name within profile rank", axis, path + ".axis is outside the profile rank"};
		return false;
	}
	const std::string *raw_side = require_string(*side_value, path + ".side", error);
	if (raw_side == nullptr) {
		return false;
	}
	const std::string side = trim(*raw_side);
	if (!contains(generated::VALID_BOUNDARY_SIDES, std::string_view(side))) {
		error = {"invalid_side", path + ".side", "- or +", side, path + ".side must be a contract boundary side"};
		return false;
	}
	result.axis = std::distance(generated::AXIS_NAMES.begin(), axis_found);
	result.side = side;
	return true;
}

bool decode_transform(
		const TransportValue &value,
		const std::string &path,
		std::int64_t rank,
		TopologyTransportTransform &result,
		TopologyTransportError &error) {
	const TransportValue::Object *object = require_object(value, path, error);
	if (object == nullptr || !require_exact_fields(*object, {"permutation", "signs"}, path, error)) {
		return false;
	}
	if (!decode_integer_array(object->at("permutation"), path + ".permutation", result.permutation, error) ||
			!decode_integer_array(object->at("signs"), path + ".signs", result.signs, error)) {
		return false;
	}
	const std::size_t tangent_rank = static_cast<std::size_t>(rank - 1);
	if (result.permutation.size() != tangent_rank) {
		error = {"invalid_permutation", path + ".permutation", "one entry per tangent axis", std::to_string(result.permutation.size()), path + ".permutation length must equal rank minus one"};
		return false;
	}
	if (result.signs.size() != tangent_rank) {
		error = {"invalid_sign", path + ".signs", "one sign per tangent axis", std::to_string(result.signs.size()), path + ".signs length must equal rank minus one"};
		return false;
	}
	std::set<std::int64_t> seen;
	for (std::size_t index = 0; index < tangent_rank; ++index) {
		const std::int64_t permutation = result.permutation[index];
		if (permutation < 0 || permutation >= rank - 1 || !seen.insert(permutation).second) {
			error = {"invalid_permutation", path + ".permutation[" + std::to_string(index) + "]", "unique tangent-axis index", std::to_string(permutation), path + ".permutation must be a complete permutation"};
			return false;
		}
		const std::int64_t sign = result.signs[index];
		if (!contains(generated::VALID_TRANSFORM_SIGNS, sign)) {
			error = {"invalid_sign", path + ".signs[" + std::to_string(index) + "]", "-1 or 1", std::to_string(sign), path + ".signs must contain contract signs"};
			return false;
		}
	}
	return true;
}

std::vector<std::int64_t> boundary_extents(
		const std::vector<std::int64_t> &dimensions,
		std::int64_t axis) {
	std::vector<std::int64_t> result;
	for (std::size_t index = 0; index < dimensions.size(); ++index) {
		if (static_cast<std::int64_t>(index) != axis) {
			result.push_back(dimensions[index]);
		}
	}
	return result;
}

bool same_boundary(const TopologyTransportBoundary &left, const TopologyTransportBoundary &right) {
	return left.axis == right.axis && left.side == right.side;
}

std::string boundary_key(const TopologyTransportBoundary &boundary) {
	return std::to_string(boundary.axis) + boundary.side;
}

} // namespace

std::string transport_value_category(const TransportValue &value) {
	if (std::holds_alternative<std::monostate>(value.value)) return "null";
	if (std::holds_alternative<bool>(value.value)) return "boolean";
	if (std::holds_alternative<std::int64_t>(value.value)) return "integer";
	if (std::holds_alternative<double>(value.value)) return "float";
	if (std::holds_alternative<std::string>(value.value)) return "string";
	if (std::holds_alternative<TransportValue::Array>(value.value)) return "array";
	if (std::holds_alternative<TransportValue::Object>(value.value)) return "object";
	return std::get<UnsupportedTransportValue>(value.value).category;
}

TopologyTransportResult<TopologyTransportProfile> decode_topology_transport_profile(
		const TransportValue &payload) {
	TopologyTransportError error;
	const auto *object = require_object(payload, "profile", error);
	if (object == nullptr || !require_exact_fields(*object, {"contract_version", "dimensions", "rank", "seams"}, "profile", error)) {
		return failure<TopologyTransportProfile>(error.code, error.path, error.expected, error.actual, error.message);
	}
	TopologyTransportProfile profile;
	const std::int64_t *version = require_integer(object->at("contract_version"), "profile.contract_version", error);
	if (version == nullptr) {
		return failure<TopologyTransportProfile>(error.code, error.path, error.expected, error.actual, error.message);
	}
	if (*version != generated::CONTRACT_VERSION) {
		return failure<TopologyTransportProfile>("contract_version_mismatch", "profile.contract_version", std::to_string(generated::CONTRACT_VERSION), std::to_string(*version), "profile.contract_version is unsupported");
	}
	profile.contract_version = *version;
	const std::int64_t *rank = require_integer(object->at("rank"), "profile.rank", error);
	if (rank == nullptr) {
		return failure<TopologyTransportProfile>(error.code, error.path, error.expected, error.actual, error.message);
	}
	if (*rank < generated::MINIMUM_RANK || *rank > generated::MAXIMUM_RANK) {
		return failure<TopologyTransportProfile>("invalid_rank", "profile.rank", "contract rank range", std::to_string(*rank), "profile.rank is outside the contract range");
	}
	profile.rank = *rank;
	if (!decode_integer_array(object->at("dimensions"), "profile.dimensions", profile.dimensions, error)) {
		return failure<TopologyTransportProfile>(error.code, error.path, error.expected, error.actual, error.message);
	}
	if (profile.dimensions.size() != static_cast<std::size_t>(profile.rank)) {
		return failure<TopologyTransportProfile>("invalid_dimension_count", "profile.dimensions", "one dimension per rank axis", std::to_string(profile.dimensions.size()), "profile.dimensions length must equal rank");
	}
	std::int64_t volume = 1;
	for (std::size_t index = 0; index < profile.dimensions.size(); ++index) {
		const std::int64_t dimension = profile.dimensions[index];
		const std::string path = "profile.dimensions[" + std::to_string(index) + "]";
		if (dimension < generated::MINIMUM_AXIS_LENGTH || dimension > generated::MAXIMUM_AXIS_LENGTH) {
			return failure<TopologyTransportProfile>("out_of_range", path, "contract axis-length range", std::to_string(dimension), path + " is outside the contract axis-length range");
		}
		if (volume > generated::MAXIMUM_INDEXABLE_VOLUME / dimension) {
			return failure<TopologyTransportProfile>("volume_overflow", "profile.dimensions", "product at most maximum indexable volume", "product exceeds maximum", "profile.dimensions product exceeds the contract maximum");
		}
		volume *= dimension;
	}
	const TransportValue::Array *seams = require_array(object->at("seams"), "profile.seams", error);
	if (seams == nullptr) {
		return failure<TopologyTransportProfile>(error.code, error.path, error.expected, error.actual, error.message);
	}
	std::set<std::string> ids;
	std::set<std::string> active_boundaries;
	for (std::size_t index = 0; index < seams->size(); ++index) {
		const std::string path = "profile.seams[" + std::to_string(index) + "]";
		const TransportValue::Object *row = require_object((*seams)[index], path, error);
		if (row == nullptr || !require_exact_fields(*row, {"enabled", "id", "source", "target", "transform"}, path, error)) {
			return failure<TopologyTransportProfile>(error.code, error.path, error.expected, error.actual, error.message);
		}
		TopologyTransportSeam seam;
		const std::string *raw_id = require_string(row->at("id"), path + ".id", error);
		if (raw_id == nullptr) {
			return failure<TopologyTransportProfile>(error.code, error.path, error.expected, error.actual, error.message);
		}
		seam.id = trim(*raw_id);
		if (seam.id.empty() || !ids.insert(seam.id).second) {
			return failure<TopologyTransportProfile>("unknown_required_value", path + ".id", "unique non-empty string", seam.id.empty() ? "empty" : seam.id, path + ".id must be unique and non-empty");
		}
		if (!decode_boundary(row->at("source"), path + ".source", profile.rank, seam.source, error) ||
				!decode_boundary(row->at("target"), path + ".target", profile.rank, seam.target, error) ||
				!decode_transform(row->at("transform"), path + ".transform", profile.rank, seam.transform, error)) {
			return failure<TopologyTransportProfile>(error.code, error.path, error.expected, error.actual, error.message);
		}
		const bool *enabled = require_boolean(row->at("enabled"), path + ".enabled", error);
		if (enabled == nullptr) {
			return failure<TopologyTransportProfile>(error.code, error.path, error.expected, error.actual, error.message);
		}
		seam.enabled = *enabled;
		if (same_boundary(seam.source, seam.target)) {
			return failure<TopologyTransportProfile>("invalid_side", path + ".target", "boundary distinct from source", boundary_key(seam.target), path + " cannot identify a boundary with itself");
		}
		const auto source_extents = boundary_extents(profile.dimensions, seam.source.axis);
		const auto target_extents = boundary_extents(profile.dimensions, seam.target.axis);
		for (std::size_t source_index = 0; source_index < seam.transform.permutation.size(); ++source_index) {
			const std::size_t target_index = static_cast<std::size_t>(seam.transform.permutation[source_index]);
			if (source_extents[source_index] != target_extents[target_index]) {
				return failure<TopologyTransportProfile>("invalid_permutation", path + ".transform.permutation[" + std::to_string(source_index) + "]", "extent-compatible tangent mapping", std::to_string(seam.transform.permutation[source_index]), path + ".transform does not preserve boundary extents");
			}
		}
		if (seam.enabled &&
				(!active_boundaries.insert(boundary_key(seam.source)).second ||
				 !active_boundaries.insert(boundary_key(seam.target)).second)) {
			return failure<TopologyTransportProfile>("unknown_required_value", path, "unused active boundaries", "duplicate active boundary", path + " reuses an active boundary");
		}
		profile.seams.push_back(std::move(seam));
	}
	return success(std::move(profile));
}

TopologyTransportResult<TopologyTransportQuery> decode_topology_transport_query(
		const TransportValue &payload,
		const TopologyTransportProfile &profile) {
	TopologyTransportError error;
	const auto *object = require_object(payload, "query", error);
	if (object == nullptr || !require_exact_fields(*object, {"axis", "coordinate", "delta", "dimensions"}, "query", error)) {
		return failure<TopologyTransportQuery>(error.code, error.path, error.expected, error.actual, error.message);
	}
	TopologyTransportQuery query;
	if (!decode_integer_array(object->at("dimensions"), "query.dimensions", query.dimensions, error) ||
			!decode_integer_array(object->at("coordinate"), "query.coordinate", query.coordinate, error)) {
		return failure<TopologyTransportQuery>(error.code, error.path, error.expected, error.actual, error.message);
	}
	if (query.dimensions.size() != static_cast<std::size_t>(profile.rank)) {
		return failure<TopologyTransportQuery>("invalid_dimension_count", "query.dimensions", "one dimension per profile rank axis", std::to_string(query.dimensions.size()), "query.dimensions length must equal profile rank");
	}
	if (query.dimensions != profile.dimensions) {
		return failure<TopologyTransportQuery>("unknown_required_value", "query.dimensions", "dimensions equal to validated profile", "different dimensions", "query.dimensions must equal profile.dimensions");
	}
	if (query.coordinate.size() != static_cast<std::size_t>(profile.rank)) {
		return failure<TopologyTransportQuery>("coordinate_rank_mismatch", "query.coordinate", "one coordinate per profile rank axis", std::to_string(query.coordinate.size()), "query.coordinate length must equal profile rank");
	}
	for (std::size_t index = 0; index < query.coordinate.size(); ++index) {
		if (query.coordinate[index] < 0 || query.coordinate[index] >= query.dimensions[index]) {
			const std::string path = "query.coordinate[" + std::to_string(index) + "]";
			return failure<TopologyTransportQuery>("coordinate_out_of_bounds", path, "coordinate inside board dimensions", std::to_string(query.coordinate[index]), path + " is outside the board");
		}
	}
	const std::int64_t *axis = require_integer(object->at("axis"), "query.axis", error);
	if (axis == nullptr) {
		return failure<TopologyTransportQuery>(error.code, error.path, error.expected, error.actual, error.message);
	}
	if (*axis < 0 || *axis >= profile.rank) {
		return failure<TopologyTransportQuery>("invalid_axis", "query.axis", "axis index within profile rank", std::to_string(*axis), "query.axis is outside the profile rank");
	}
	query.axis = *axis;
	const std::int64_t *delta = require_integer(object->at("delta"), "query.delta", error);
	if (delta == nullptr) {
		return failure<TopologyTransportQuery>(error.code, error.path, error.expected, error.actual, error.message);
	}
	if (!contains(generated::VALID_MOVEMENT_DELTAS, *delta)) {
		return failure<TopologyTransportQuery>("invalid_delta", "query.delta", "-1 or 1", std::to_string(*delta), "query.delta must be a contract unit movement delta");
	}
	query.delta = *delta;
	return success(std::move(query));
}

TopologyQueryProfile topology_query_profile_from_transport(
		const TopologyTransportProfile &profile) {
	TopologyQueryProfile result;
	result.dimension = static_cast<int>(profile.rank);
	for (const TopologyTransportSeam &seam : profile.seams) {
		BoundaryQueryTransform transform;
		for (const std::int64_t value : seam.transform.permutation) transform.permutation.push_back(static_cast<int>(value));
		for (const std::int64_t value : seam.transform.signs) transform.signs.push_back(static_cast<int>(value));
		result.gluings.push_back(GluingQueryDescriptor{
				seam.id,
				BoundaryQueryRef{static_cast<int>(profile.rank), static_cast<int>(seam.source.axis), seam.source.side == "-" ? -1 : 1},
				BoundaryQueryRef{static_cast<int>(profile.rank), static_cast<int>(seam.target.axis), seam.target.side == "-" ? -1 : 1},
				std::move(transform),
				seam.enabled});
	}
	return result;
}

TopologyCellStepQueryResult resolve_topology_transport_query(
		const TopologyTransportProfile &profile,
		const TopologyTransportQuery &query) {
	std::vector<int> dimensions;
	CoordND coordinate;
	for (const std::int64_t value : query.dimensions) dimensions.push_back(static_cast<int>(value));
	for (const std::int64_t value : query.coordinate) coordinate.values.push_back(static_cast<int>(value));
	return resolve_topology_cell_step_query(
			topology_query_profile_from_transport(profile),
			BoardShapeND{std::move(dimensions)},
			coordinate,
			MoveStepQuery{static_cast<int>(query.axis), static_cast<int>(query.delta)});
}

} // namespace tet4d::core

File: src/tet4d/engine/topology_explorer/topology_transport.py
from __future__ import annotations

from collections.abc import Mapping
from dataclasses import dataclass

from tet4d.generated.topology_contract_v1 import (
    AXIS_NAMES,
    CONTRACT_VERSION,
    MAXIMUM_AXIS_LENGTH,
    MAXIMUM_INDEXABLE_VOLUME,
    MAXIMUM_RANK,
    MINIMUM_AXIS_LENGTH,
    MINIMUM_RANK,
    VALID_BOUNDARY_SIDES,
    VALID_MOVEMENT_DELTAS,
    VALID_TRANSFORM_SIGNS,
)


@dataclass(frozen=True)
class TopologyTransportError(ValueError):
    code: str
    path: str
    expected: str
    actual: str
    message: str

    def __str__(self) -> str:
        return self.message


def _category(value: object) -> str:
    if value is None:
        return "null"
    if type(value) is bool:
        return "boolean"
    if type(value) is int:
        return "integer"
    if type(value) is float:
        return "float"
    if type(value) is str:
        return "string"
    if type(value) is list:
        return "array"
    if type(value) is dict:
        return "object"
    return type(value).__name__


def _fail(code: str, path: str, expected: str, actual: str, message: str) -> None:
    raise TopologyTransportError(code, path, expected, actual, message)


def _object(value: object, path: str) -> dict[str, object]:
    if type(value) is not dict:
        _fail(
            "wrong_type", path, "object", _category(value), f"{path} must be an object"
        )
    assert isinstance(value, dict)
    if any(type(key) is not str for key in value):
        _fail(
            "wrong_type",
            path,
            "object with string keys",
            "non-string key",
            f"{path} keys must be strings",
        )
    return value


def _array(value: object, path: str) -> list[object]:
    if type(value) is not list:
        _fail("wrong_type", path, "array", _category(value), f"{path} must be an array")
    assert isinstance(value, list)
    return value


def _exact_fields(
    value: Mapping[str, object], fields: frozenset[str], path: str
) -> None:
    missing = sorted(fields.difference(value))
    if missing:
        field = missing[0]
        _fail(
            "missing_field",
            f"{path}.{field}",
            "present field",
            "missing",
            f"{path}.{field} is required",
        )
    unknown = sorted(set(value).difference(fields))
    if unknown:
        field = unknown[0]
        _fail(
            "unknown_required_value",
            f"{path}.{field}",
            "known field",
            "unknown field",
            f"{path}.{field} is not supported",
        )


def _integer(value: object, path: str) -> int:
    if type(value) is not int:
        _fail(
            "wrong_type",
            path,
            "integer",
            _category(value),
            f"{path} must be an integer",
        )
    assert isinstance(value, int)
    return value


def _boolean(value: object, path: str) -> bool:
    if type(value) is not bool:
        _fail(
            "wrong_type", path, "boolean", _category(value), f"{path} must be a boolean"
        )
    assert isinstance(value, bool)
    return value


def _string(value: object, path: str) -> str:
    if type(value) is not str:
        _fail(
            "wrong_type", path, "string", _category(value), f"{path} must be a string"
        )
    assert isinstance(value, str)
    return value


def _integer_array(value: object, path: str) -> list[int]:
    return [
        _integer(item, f"{path}[{index}]")
        for index, item in enumerate(_array(value, path))
    ]


def _boundary(value: object, path: str, rank: int) -> dict[str, object]:
    row = _object(value, path)
    _exact_fields(row, frozenset(("axis", "side")), path)
    axis_name = _string(row["axis"], f"{path}.axis").strip().lower()
    if axis_name not in AXIS_NAMES[:rank]:
        _fail(
            "invalid_axis",
            f"{path}.axis",
            "axis name within profile rank",
            axis_name,
            f"{path}.axis is outside the profile rank",
        )
    side = _string(row["side"], f"{path}.side").strip()
    if side not in VALID_BOUNDARY_SIDES:
        _fail(
            "invalid_side",
            f"{path}.side",
            "- or +",
            side,
            f"{path}.side must be a contract boundary side",
        )
    return {"axis": AXIS_NAMES.index(axis_name), "side": side}


def _transform(value: object, path: str, rank: int) -> dict[str, list[int]]:
    row = _object(value, path)
    _exact_fields(row, frozenset(("permutation", "signs")), path)
    permutation = _integer_array(row["permutation"], f"{path}.permutation")
    signs = _integer_array(row["signs"], f"{path}.signs")
    tangent_rank = rank - 1
    if len(permutation) != tangent_rank:
        _fail(
            "invalid_permutation",
            f"{path}.permutation",
            "one entry per tangent axis",
            str(len(permutation)),
            f"{path}.permutation length must equal rank minus one",
        )
    if len(signs) != tangent_rank:
        _fail(
            "invalid_sign",
            f"{path}.signs",
            "one sign per tangent axis",
            str(len(signs)),
            f"{path}.signs length must equal rank minus one",
        )
    seen: set[int] = set()
    for index, item in enumerate(permutation):
        if item < 0 or item >= tangent_rank or item in seen:
            _fail(
                "invalid_permutation",
                f"{path}.permutation[{index}]",
                "unique tangent-axis index",
                str(item),
                f"{path}.permutation must be a complete permutation",
            )
        seen.add(item)
        if signs[index] not in VALID_TRANSFORM_SIGNS:
            _fail(
                "invalid_sign",
                f"{path}.signs[{index}]",
                "-1 or 1",
                str(signs[index]),
                f"{path}.signs must contain contract signs",
            )
    return {"permutation": permutation, "signs": signs}


def _boundary_extents(dimensions: list[int], axis: int) -> list[int]:
    return [value for index, value in enumerate(dimensions) if index != axis]


def _dimensions(value: object, rank: int) -> list[int]:
    dimensions = _integer_array(value, "profile.dimensions")
    if len(dimensions) != rank:
        _fail(
            "invalid_dimension_count",
            "profile.dimensions",
            "one dimension per rank axis",
            str(len(dimensions)),
            "profile.dimensions length must equal rank",
        )
    volume = 1
    for index, dimension in enumerate(dimensions):
        path = f"profile.dimensions[{index}]"
        if dimension < MINIMUM_AXIS_LENGTH or dimension > MAXIMUM_AXIS_LENGTH:
            _fail(
                "out_of_range",
                path,
                "contract axis-length range",
                str(dimension),
                f"{path} is outside the contract axis-length range",
            )
        if volume > MAXIMUM_INDEXABLE_VOLUME // dimension:
            _fail(
                "volume_overflow",
                "profile.dimensions",
                "product at most maximum indexable volume",
                "product exceeds maximum",
                "profile.dimensions product exceeds the contract maximum",
            )
        volume *= dimension
    return dimensions


def _seam(
    value: object,
    index: int,
    rank: int,
    dimensions: list[int],
    ids: set[str],
    active_boundaries: set[tuple[int, str]],
) -> dict[str, object]:
    path = f"profile.seams[{index}]"
    row = _object(value, path)
    _exact_fields(
        row,
        frozenset(("id", "source", "target", "transform", "enabled")),
        path,
    )
    seam_id = _string(row["id"], f"{path}.id").strip()
    if not seam_id or seam_id in ids:
        _fail(
            "unknown_required_value",
            f"{path}.id",
            "unique non-empty string",
            "empty" if not seam_id else seam_id,
            f"{path}.id must be unique and non-empty",
        )
    ids.add(seam_id)
    source = _boundary(row["source"], f"{path}.source", rank)
    target = _boundary(row["target"], f"{path}.target", rank)
    transform = _transform(row["transform"], f"{path}.transform", rank)
    enabled = _boolean(row["enabled"], f"{path}.enabled")
    source_key = (int(source["axis"]), str(source["side"]))
    target_key = (int(target["axis"]), str(target["side"]))
    if source_key == target_key:
        _fail(
            "invalid_side",
            f"{path}.target",
            "boundary distinct from source",
            f"{target_key[0]}{target_key[1]}",
            f"{path} cannot identify a boundary with itself",
        )
    source_extents = _boundary_extents(dimensions, source_key[0])
    target_extents = _boundary_extents(dimensions, target_key[0])
    permutation = transform["permutation"]
    for source_index, target_index in enumerate(permutation):
        if source_extents[source_index] != target_extents[target_index]:
            _fail(
                "invalid_permutation",
                f"{path}.transform.permutation[{source_index}]",
                "extent-compatible tangent mapping",
                str(target_index),
                f"{path}.transform does not preserve boundary extents",
            )
    if enabled and (source_key in active_boundaries or target_key in active_boundaries):
        _fail(
            "unknown_required_value",
            path,
            "unused active boundaries",
            "duplicate active boundary",
            f"{path} reuses an active boundary",
        )
    if enabled:
        active_boundaries.update((source_key, target_key))
    return {
        "id": seam_id,
        "source": source,
        "target": target,
        "transform": transform,
        "enabled": enabled,
    }


def _seams(value: object, rank: int, dimensions: list[int]) -> list[dict[str, object]]:
    ids: set[str] = set()
    active_boundaries: set[tuple[int, str]] = set()
    return [
        _seam(item, index, rank, dimensions, ids, active_boundaries)
        for index, item in enumerate(_array(value, "profile.seams"))
    ]


def validate_topology_transport_profile(payload: object) -> dict[str, object]:
    profile = _object(payload, "profile")
    _exact_fields(
        profile,
        frozenset(("contract_version", "rank", "dimensions", "seams")),
        "profile",
    )
    version = _integer(profile["contract_version"], "profile.contract_version")
    if version != CONTRACT_VERSION:
        _fail(
            "contract_version_mismatch",
            "profile.contract_version",
            str(CONTRACT_VERSION),
            str(version),
            "profile.contract_version is unsupported",
        )
    rank = _integer(profile["rank"], "profile.rank")
    if rank < MINIMUM_RANK or rank > MAXIMUM_RANK:
        _fail(
            "invalid_rank",
            "profile.rank",
            "contract rank range",
            str(rank),
            "profile.rank is outside the contract range",
        )
    dimensions = _dimensions(profile["dimensions"], rank)
    seams = _seams(profile["seams"], rank, dimensions)
    return {
        "contract_version": version,
        "rank": rank,
        "dimensions": dimensions,
        "seams": seams,
    }


def validate_topology_transport_query(
    payload: object, profile: Mapping[str, object]
) -> dict[str, object]:
    query = _object(payload, "query")
    _exact_fields(
        query,
        frozenset(("dimensions", "coordinate", "axis", "delta")),
        "query",
    )
    rank = int(profile["rank"])
    profile_dimensions = list(profile["dimensions"])
    dimensions = _integer_array(query["dimensions"], "query.dimensions")
    coordinate = _integer_array(query["coordinate"], "query.coordinate")
    if len(dimensions) != rank:
        _fail(
            "invalid_dimension_count",
            "query.dimensions",
            "one dimension per profile rank axis",
            str(len(dimensions)),
            "query.dimensions length must equal profile rank",
        )
    if dimensions != profile_dimensions:
        _fail(
            "unknown_required_value",
            "query.dimensions",
            "dimensions equal to validated profile",
            "different dimensions",
            "query.dimensions must equal profile.dimensions",
        )
    if len(coordinate) != rank:
        _fail(
            "coordinate_rank_mismatch",
            "query.coordinate",
            "one coordinate per profile rank axis",
            str(len(coordinate)),
            "query.coordinate length must equal profile rank",
        )
    for index, item in enumerate(coordinate):
        if item < 0 or item >= dimensions[index]:
            path = f"query.coordinate[{index}]"
            _fail(
                "coordinate_out_of_bounds",
                path,
                "coordinate inside board dimensions",
                str(item),
                f"{path} is outside the board",
            )
    axis = _integer(query["axis"], "query.axis")
    if axis < 0 or axis >= rank:
        _fail(
            "invalid_axis",
            "query.axis",
            "axis index within profile rank",
            str(axis),
            "query.axis is outside the profile rank",
        )
    delta = _integer(query["delta"], "query.delta")
    if delta not in VALID_MOVEMENT_DELTAS:
        _fail(
            "invalid_delta",
            "query.delta",
            "-1 or 1",
            str(delta),
            "query.delta must be a contract unit movement delta",
        )
    return {
        "dimensions": dimensions,
        "coordinate": coordinate,
        "axis": axis,
        "delta": delta,
    }


__all__ = [
    "TopologyTransportError",
    "validate_topology_transport_profile",
    "validate_topology_transport_query",
]
